src/units/unit.cpp

```diff
@@ -1408,6 +1408,138 @@ void unit::remove_ability_by_id(const std::string& ability)
 	}
 }
 
+static bool bool_matches_if_present(const config& filter, const config& cfg, const std::string& attribute, bool def)
+{
+	if(filter[attribute].empty()) {
+		return true;
+	}
+
+	return filter[attribute].to_bool() == cfg[attribute].to_bool(def);
+}
+
+static bool string_matches_if_present(const config& filter, const config& cfg, const std::string& attribute, const std::string& def)
+{
+	if(filter[attribute].empty()) {
+		return true;
+	}
+
+	const std::vector<std::string> filter_attribute = utils::split(filter[attribute]);
+	return ( std::find(filter_attribute.begin(), filter_attribute.end(), cfg[attribute].str(def)) != filter_attribute.end() );
+}
+
+static bool type_value_if_present(const config& filter, const config& cfg)
+{
+	if(filter["type_value"].empty()) {
+		return true;
+	}
+
+	std::string cfg_type_value;
+	const std::vector<std::string> filter_attribute = utils::split(filter["type_value"]);
+	if(!cfg["value"].empty()){
+		cfg_type_value ="value";
+	} else if(!cfg["add"].empty()){
+		cfg_type_value ="add";
+	} else if(!cfg["sub"].empty()){
+		cfg_type_value ="sub";
+	} else if(!cfg["multiply"].empty()){
+		cfg_type_value ="multiply";
+	} else if(!cfg["divide"].empty()){
+		cfg_type_value ="divide";
+	}
+	return ( std::find(filter_attribute.begin(), filter_attribute.end(), cfg_type_value) != filter_attribute.end() );
+}
+
+static bool matches_ability_filter(const config & cfg, const std::string& tag_name, const config & filter, bool ignore_tag_name)
+{
+	if(!bool_matches_if_present(filter, cfg, "affect_self", true))
+		return false;
+
+	if(!bool_matches_if_present(filter, cfg, "affect_allies", true))
+		return false;
+
+	if(!bool_matches_if_present(filter, cfg, "affect_enemies", false))
+		return false;
+
+	if(!bool_matches_if_present(filter, cfg, "cumulative", false))
+		return false;
+
+	if(!ignore_tag_name){
+		const std::vector<std::string> filter_type = utils::split(filter["tag_name"]);
+		if ( !filter_type.empty() && std::find(filter_type.begin(), filter_type.end(), tag_name) == filter_type.end() ){
+			return false;
+		}
+	}
+	if(!string_matches_if_present(filter, cfg, "id", ""))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "apply_to", "self"))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "overwrite_specials", "none"))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "active_on", "both"))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "value", ""))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "add", ""))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "sub", ""))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "multiply", ""))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "divide", ""))
```

Comment:
Values should be matched as a number, not a string. Matching as `double` should be sufficient.

Also, I feel that a filter on `add=-2` should match an ability with `sub=2` and vice versa. Similarly, a filter with `divide=2` should match an ability with `multiply=0.5` and vice versa.

Reply:
Except that the use of string allows to filter a list of values like value=10,20, etc.. for each attribute instead of having to write
value=10
[or]
value=20
[or]
etc..
that must be why neither int nor double are used in the encoding of [filter_weapon] in the units/attack_type.cpp file.
As for the equivalence add=2 sub=-2, I don't believe it should be coded in the source code whereas a wml developer can code
add=2
[gold]
sub=-2

if he deems it necessary, but this should not be systematic, given that I consider, for my part, that the filtering of the writing of the operation is as important as that of the operation itself for a minimum clarity for the wml developer and thus avoid confusing it and less risk of bugs than if I try to encode an equivalence that I consider useless.

Reply:
> Except that the use of string allows to filter a list of values like value=10,20, etc.. for each attribute

No, that's still wrong. If you write `value=10.0,20.0` your code won't work. There's also no way to encode "match a number between 1 and 12". Supporting a comma-separated list has nothing to do with the fact that you're interpreting it as a string. You can still support a comma-separated list even while interpreting it as a number, using the function `parse_ranges`. This also adds support for syntax like `1-12`.



> As for the equivalence add=2 sub=-2, I don't believe it should be coded in the source code whereas a wml developer can code add=2 [or] sub=-2
> 
> if he deems it necessary, but this should not be systematic, given that I consider, for my part, that the filtering of the writing of the operation is as important as that of the operation itself for a minimum clarity for the wml developer and thus avoid confusing it and less risk of bugs than if I try to encode an equivalence that I consider useless.

I don't really understand why you'd want to not consider equivalent things to be equivalent. An ability that applies `add=-2` does the exact same thing as an ability that applies `sub=2`. An ability that applies `sub=-2` does the exact same thing as an ability that applies `add=2`. An ability that applies `multiply=0.5` does the exact same thing as an ability that applies `divide=2`. Why would the way the developer chooses to express it affect whether the filter matches?

You say it would be confusing to code this equivalence. Is it not more confusing if your filter with `divide=2` _doesn't always_ match an ability that halves the value? Is it not more confusing if your filter with `sub=2` _doesn't always_ match an ability that subtracts 2 from the value?

Besides which, you've already coded a way to match `add` but not `sub` if you really want to, using `type_value=add`. So, if someone really wants a filter that matches `add=-2` but does not match `sub=2`, they can just write:

```
add=-2
type_value=add
```

…right?

---

I think there is also another issue lurking here, which is the fact that `value` and related keys can also be a formula instead of a number. I think this is a case that needs to be handled somehow… but I'm not sure what is the best way to do so.

Reply:
it's done in https://github.com/wesnoth/wesnoth/pull/7168/commits/2f36bd118dda5e27ebbc5e6888bb71ab0c315196 commit

Reply:
i removed commit because i feels what i causes more issue, i i would what this PR finally merged before to resolve that

Reply:
@CelticMinstrel i re pushed new version of conversion in https://github.com/wesnoth/wesnoth/pull/7168/commits/beee74e564f5cac3be3c571419a7f755cf2d9ca8 and re-ordered commit (titles with [1.17] must be showed inmaster after merging)

Reply:
Having looked into [the implementation of `effect::effect`](https://github.com/wesnoth/wesnoth/pull/7050#issuecomment-1365145483), I'm thinking there should only be `[filter_ability]add=` which matches both `[ability]add=` and `[ability]sub=`, with no corresponding `[filter_ability]sub=`.

Also based on the implementation, `add` should only support integers.

Reply:
It doesn't seem terrible to support both `add` and `sub` though as near-synonyms.

Making them be integers only seems fine.